aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -376,6 +376,93 @@ Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
   }
 }
 
+template <typename T, bool upper>
+struct BatchTensorTriOp {
+  BatchTensorTriOp(T* start_, int64_t stride_batch_, int64_t stride_row_, int64_t stride_col_, int64_t k_)
+    : start(start_), stride_batch(stride_batch_), stride_row(stride_row_), stride_col(stride_col_), k(k_) {}
+
+  __device__ __forceinline__ int mask(T* result) {
+    ptrdiff_t n = result - start;
+    bool stride_cond = stride_row > stride_col;
+    int64_t stride_max = stride_cond ? stride_row : stride_col;
+    int64_t stride_min = stride_cond ? stride_col : stride_row;
+
+    if (stride_batch > stride_max) {
+      n = n % stride_batch;
+    } else if (stride_batch > stride_min) {
+      n = n - n % stride_max + n % stride_batch;
+    }
+
+    int64_t row, col;
+    if (stride_cond) {
+      row = (int64_t) (n / stride_row);
+      col = (int64_t) ((n % stride_row) / stride_col);
+    } else {
+      row = (int64_t) ((n % stride_col) / stride_row);
+      col = (int64_t) (n / stride_col);
+    }
+    return upper ? (col - row >= k) : (col - row <= k);
+  }
+
+  __device__ __forceinline__ void operator()(T* result, T* self) {
```

Comment:
Does changing this to ` __device__ __forceinline__ void operator()(T& result, T& self) {`

(and the corresponding operator()(T* v) to take a reference) help?

Reply:
The operator passed to CUDA_tensor_apply2 should take in references, not pointers, so I think this should fix your compilation error: https://github.com/pytorch/pytorch/blob/bb9b7de831d25ab40d64aca6069d8166ee5d92f7/aten/src/ATen/native/cuda/Activation.cu#L21-L29

Reply:
Thank you, I’ll check if this works out locally.

Reply:
No, this doesn't work unfortunately. The leading error in this case is:

```
/media/vishwak/Official/Projects/pytorch/aten/src/ATen/cuda/CUDAApplyUtils.cuh(331): error: no instance of overloaded function "at::native::BatchTensorTriOp<T, upper>::operator() [with T=uint8_t, upper=false]" matches the argument list and object (the object has type qualifiers that prevent a match)
            argument types are: (uint8_t, uint8_t)
            object type is: const at::native::BatchTensorTriOp<uint8_t, false>
          detected during:
            instantiation of "void at::cuda::<unnamed>::ApplyOp2<Op, scalar1, scalar2, IndexType, ADims, BDims, 0, Offset>::apply(at::cuda::detail::TensorInfo<scalar1, IndexType> &, at::cuda::detail::TensorInfo<scalar2, IndexType> &, const Op &, int, IndexType, Offset, Offset) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, Offset=const unsigned int]" 
(310): here
            instantiation of "void at::cuda::<unnamed>::ApplyOp2<Op, scalar1, scalar2, IndexType, ADims, BDims, remaining_steps, Offsets...>::apply(at::cuda::detail::TensorInfo<scalar1, IndexType> &, at::cuda::detail::TensorInfo<scalar2, IndexType> &, const Op &, int, IndexType, Offsets..., Offsets...) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, remaining_steps=1, Offsets=<>]" 
(369): here
            instantiation of "void at::cuda::<unnamed>::kernelPointwiseApply2<Op,scalar1,scalar2,IndexType,ADims,BDims,step>(at::cuda::detail::TensorInfo<scalar1, IndexType>, at::cuda::detail::TensorInfo<scalar2, IndexType>, IndexType, Op) [with Op=at::native::BatchTensorTriOp<uint8_t, false>, scalar1=uint8_t, scalar2=uint8_t, IndexType=unsigned int, ADims=1, BDims=1, step=1]" 
(888): here
            instantiation of "__nv_bool at::cuda::CUDA_tensor_apply2<scalar1,scalar2,step,Op>(at::Tensor, at::Tensor, Op, at::cuda::TensorArgType, at::cuda::TensorArgType) [with scalar1=uint8_t, scalar2=uint8_t, step=1, Op=at::native::BatchTensorTriOp<uint8_t, false>]" 
(940): here
            instantiation of "__nv_bool at::cuda::CUDA_tensor_apply2<scalar1,scalar2,Op>(at::Tensor, at::Tensor, Op, at::cuda::TensorArgType, at::cuda::TensorArgType) [with scalar1=uint8_t, scalar2=uint8_t, Op=at::native::BatchTensorTriOp<uint8_t, false>]" 
/media/vishwak/Official/Projects/pytorch/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu(430): here
            instantiation of "void at::native::apply_triu_tril<scalar_t,inplace,upper>(at::Tensor &, const at::Tensor &, int64_t) [with scalar_t=uint8_t, inplace=true, upper=false]" 
/media/vishwak/Official/Projects/pytorch/aten/src/ATen/native/cuda/BatchLinearAlgebra.cu(437): here
```

This reminded me that I had already tried this earlier, and this didn't work. In fact, this was how @t-vi had kept it in his original PR.

Reply:
I don't know much about it except that my patch worked at the time of submission, but so the argument types it tries to call with are uint8_t, uint8_t and so if you declare that the first needs to be a const, it would not work.

Reply:
I circumvented this issue by using thrust::tabulate. Thank you @zou3519 @t-vi for your suggestions, nonetheless. :)